jujugui/static/gui/src/app/components/terminal/terminal.js

```diff
@@ -0,0 +1,126 @@
+/* Copyright (C) 2017 Canonical Ltd. */
+'use strict';
+
+const React = require('react');
+const PropTypes = require('prop-types');
+const shapeup = require('shapeup');
+const XTerm = require('xterm');
+require('xterm/lib/addons/terminado/terminado');
+
+const Lightbox = require('../lightbox/lightbox');
+
+const SvgIcon = require('../svg-icon/svg-icon');
+
+/** Terminal component used to display the Juju shell. */
+class Terminal extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {opened: false};
+    this.term = null;
+    this.ws = null;
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const state = this.state;
+    if (prevState.opened && !state.opened) {
+      this.stopTerm();
+      return;
+    }
+    if (!prevState.opened && state.opened) {
+      this.startTerm();
+    }
+  }
+
+  setOpened(opened, evt) {
+    if (evt) {
+      evt.stopPropagation();
+      evt.preventDefault();
+    }
+    this.setState({opened: opened});
+  }
+
+  startTerm() {
+    const props = this.props;
+    const ws = new WebSocket(`ws://${props.address}/ws/`);
+    ws.onopen = () => {
+      ws.send(JSON.stringify({
+        operation: 'login',
+        username: props.creds.user,
+        password: props.creds.password,
+        macaroons: props.creds.macaroons
+      }));
+      ws.send(JSON.stringify({operation: 'start'}));
+    };
+    ws.onerror = err => {
+      console.error('WebSocket error:', err);
+    };
+    ws.onmessage = evt => {
+      const resp = JSON.parse(evt.data);
+      if (resp.code === 'error') {
+        console.error(resp.message);
+        return;
+      }
+      if (resp.code === 'ok' && resp.message === 'session is ready') {
+        const term = new XTerm();
+        term.terminadoAttach(ws);
+        term.open(document.getElementById('juju-shell'));
```

Comment:
instead of querying the document, this should be queried using something like `ReactDOM.findDOMNode(this).querySelector('.juju-shell__terminal-container')`